File: Book-AlgorithmsAndDataStructuresForProgrammingContests/P103StlSamples/stack.cpp
// P.104
#include <iostream>
#include <stack>
using namespace std;
int main() {
  stack<int> S;

  S.push(3);
  cout << "Push 3 -> Stack top: " << S.top() << ", size: " << S.size() << endl;
  S.push(7);
  cout << "Push 7 -> Stack top: " << S.top() << ", size: " << S.size() << endl;
  S.push(1);
  cout << "Push 1 -> Stack top: " << S.top() << ", size: " << S.size() << endl;

  S.pop();
  cout << "Pop    -> Stack top: " << S.top() << ", size: " << S.size() << endl;
  S.pop();
  cout << "Pop    -> Stack top: " << S.top() << ", size: " << S.size() << endl;

  S.push(5);
  cout << "Push 5 -> Stack top: " << S.top() << ", size: " << S.size() << endl;

  S.pop();
  cout << "Pop    -> Stack top: " << S.top() << ", size: " << S.size() << endl;

  return 0;
}

File: Book-AlgorithmsAFunctionalProgrammingApproachSecondEdition/hsorig/Stack.hs
module Stack (Stack(..),push,pop,top,emptyStack,stackEmpty) where

-- | P.87
emptyStack :: Stack a
-- | P.87
stackEmpty :: Stack a -> Bool
-- | P.87
push :: a -> Stack a -> Stack a
-- | P.87
pop :: Stack a -> Stack a
-- | P.87
top :: Stack a -> a

-- | P.88, implementation with a constructor type
data Stack a = EmptyStk | Stk a (Stack a) deriving Eq

instance (Show a) => Show (Stack a) where
  showsPrec p EmptyStk  str = showChar '-' str
  showsPrec p (Stk x s) str = shows x (showChar '|' (shows s str))

-- | P.88
emptyStack = EmptyStk

-- | P.88
stackEmpty EmptyStk = True
stackEmpty _        = False

-- | P.88
push = Stk

-- | P.88
pop EmptyStk  = error "pop from an empty stack"
pop (Stk _ s) = s

-- | P.88
top EmptyStk  = error "top from an empty stack"
top (Stk x _) = x
